infobind/cache.go

```diff
@@ -0,0 +1,224 @@
+		}
+		if chk.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record, err := decodeBindTableRow(row, fs)
+			if err != nil {
+				log.Errorf("row decode error %s", err)
+				continue
+			}
+			err = bc.appendNode(h.Ctx, record, h.Parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.LastUpdateTime) == 1 {
+				h.LastUpdateTime = record.UpdateTime
+			}
+		}
+		chk = chunk.Renew(chk, h.Ctx.GetSessionVars().MaxChunkSize)
+	}
+}
+
+// Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.Get()
+
+	length := defaultBindCacheSize
+	if bc != nil {
+		length = len(bc.Cache)
+	}
+
+	newBc := &BindCache{
+		Cache: make(map[string][]*BindData, length),
+	}
+
+	if bc != nil {
+		for hash, bindDataArr := range bc.Cache {
+			newBc.Cache[hash] = append(newBc.Cache[hash], bindDataArr...)
+		}
+	}
+
+	if fullLoad {
+		sql = fmt.Sprintf("select * from mysql.bind_info")
+	} else {
+		sql = fmt.Sprintf("select * from mysql.bind_info where update_time > \"%s\"", h.LastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, newBc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.bind.Store(newBc)
+	return nil
+}
+
+func parseSQL(sctx sessionctx.Context, parser *parser.Parser, sql string, charset string, collation string) ([]ast.StmtNode, []error, error) {
+	return parser.Parse(sql, charset, collation)
+}
+
+func decodeBindTableRow(row chunk.Row, fs []*ast.ResultField) (bindRecord, error) {
+	var value bindRecord
+
+	value.OriginalSQL = row.GetString(0)
+	value.BindSQL = row.GetString(1)
+	value.Db = row.GetString(2)
+	value.Status = row.GetInt64(3)
+	value.CreateTime = row.GetTime(4)
+	value.UpdateTime = row.GetTime(5)
+	value.Charset = row.GetString(6)
+	value.Collation = row.GetString(7)
+
+	return value, nil
+}
+
+func (b *BindCache) appendNode(sctx sessionctx.Context, value bindRecord, sparser *parser.Parser) error {
+	hash := parser.Digest(value.OriginalSQL)
+	if value.Status == 0 {
+		if bindArr, ok := b.Cache[hash]; ok {
+			if len(bindArr) == 1 {
+				if bindArr[0].Db == value.Db {
+					delete(b.Cache, hash)
+				}
+				return nil
+			}
+			for idx, v := range bindArr {
+				if v.OriginalSQL == value.OriginalSQL && v.Db == value.Db {
+					b.Cache[hash] = append(b.Cache[hash][:idx], b.Cache[hash][idx+1:]...)
+					break
+				}
+			}
+		}
+		return nil
+	}
+
+	stmtNodes, _, err := parseSQL(sctx, sparser, value.BindSQL, value.Charset, value.Collation)
+	if err != nil {
+		log.Warnf("parse error:\n%v\n%s", err, value.BindSQL)
+		return errors.Trace(err)
+	}
+
+	newNode := &BindData{
+		bindRecord: value,
+		ast:        stmtNodes[0],
+	}
+
+	if bindArr, ok := b.Cache[hash]; ok {
+		for idx, v := range bindArr {
+			if v.Db == value.Db {
+				b.Cache[hash][idx] = newNode
```

Comment:
Why directly change it?

Reply:
I have rewrite this code, please review new code.